external_parser/example_joiner.cc

```diff
@@ -0,0 +1,162 @@
+
+    joined_event info = {"joiner_timestamp", std::move(meta), std::move(data)};
+    _unjoined_examples.emplace(metadata.id()->str(), std::move(info));
+  }
+  return 0;
+}
+
+int ExampleJoiner::process_outcome(const v2::Event &event,
+                                   const v2::Metadata &metadata) {
+
+  outcome_event o_event;
+  o_event.metadata = {"client_time_utc",
+                      metadata.app_id() ? metadata.app_id()->str() : "",
+                      metadata.payload_type(), metadata.pass_probability(),
+                      metadata.encoding()};
+
+  auto outcome = v2::GetOutcomeEvent(event.payload()->data());
+
+  int index = -1;
+
+  std::cout << "outcome: value:";
+  if (outcome->value_type() == v2::OutcomeValue_literal) {
+    o_event.s_value = outcome->value_as_literal()->c_str();
+    std::cout << outcome->value_as_literal()->c_str();
+  } else if (outcome->value_type() == v2::OutcomeValue_numeric) {
+    o_event.value = outcome->value_as_numeric()->value();
+    std::cout << outcome->value_as_numeric()->value();
+  }
+
+  std::cout << " index:";
+  if (outcome->index_type() == v2::IndexValue_literal) {
+    std::cout << outcome->index_as_literal()->c_str();
+    o_event.s_index = outcome->index_as_literal()->c_str();
+    index = std::stoi(outcome->index_as_literal()->c_str());
+  } else if (outcome->index_type() == v2::IndexValue_numeric) {
+    std::cout << outcome->index_as_numeric()->index();
+    o_event.s_index = outcome->index_as_numeric()->index();
+    index = outcome->index_as_numeric()->index();
+  }
+
+  std::cout << " action-taken:" << outcome->action_taken() << std::endl;
+
+  if (_unjoined_examples.find(metadata.id()->str()) !=
+      _unjoined_examples.end()) {
+    auto &joined_event = _unjoined_examples[metadata.id()->str()];
+    joined_event.outcome_events.push_back(o_event);
+  }
+
+  return 0;
+}
+
+int ExampleJoiner::process_joined(v_array<example *> &examples) {
+  if (_event_order.empty()) {
+    return 0;
+  }
+
+  // TODO event order store events backwards
+  auto &id = _event_order[0];
+  for (auto *event : _unjoined_events[id]) {
+    auto metadata = event->meta();
+    std::cout << "id:" << metadata->id()->c_str()
+              << " type:" << v2::EnumNamePayloadType(metadata->payload_type())
+              << " payload-size:" << event->payload()->size()
+              << " encoding:" << v2::EnumNameEventEncoding(metadata->encoding())
+              << std::endl;
+
+    if (metadata->encoding() == v2::EventEncoding_Zstd) {
+      std::cout << "Decompression coming soon" << std::endl;
+    }
+
+    if (metadata->payload_type() == v2::PayloadType_Outcome) {
+      process_outcome(*event, *metadata);
+    } else {
+      process_interaction(*event, *metadata, examples);
+    }
+  }
+  // call logic that creates the reward
+  reward_calculation(_unjoined_examples[id], examples);
+  // return an empty example to signal end-of-multiline
```

Comment:
oh need to add check here that we are actually multi-line before adding an empty example